### PR TITLE
Fix insertTextAtRange

### PR DESCRIPTION
**Pull request by zhujinxuan**
## Description:
change insertTextAtRange logic to insert and then delete

## Why:
startKey can be deleted in deleteAtRange in hanging selection

### CODE REVIEW

**Comment by zhujinxuan** [comment]
Update: 
In recent commit, I switch the logic to insert and then delete; it seems that method solves the problem in a simpler way.

File: packages/slate/src/changes/at-current-range.js [hunk 1]
```diff
@@ -199,7 +199,7 @@ Changes.insertText = (change, text, marks) => {
   const { value } = change
   const { document, selection } = value
   marks = marks || selection.marks || document.getInsertMarksAtRange(selection)
-  change.insertTextAtRange(selection, text, marks)
+  change.insertTextAtRange(selection, text, marks, { normalize: true })
```
Comment (zhujinxuan) [inline packages/slate/src/changes/at-current-range.js]: Remember to remove `{normalize:true}` and add back PREF in at-range.

Comment (zhujinxuan) [reply packages/slate/src/changes/at-current-range.js]: Fixed

**Comment by ianstormtaylor** [comment]
@zhujinxuan why not use the same trick you used in the other pull request, so that we still delete first (since I think it's simpler?) but then right after the delete we properly fix the range if the `startKey` was deleted?

**Comment by zhujinxuan** [comment]
@ianstormtaylor Because considern this:
```
<paragraph><anchor/> Cat is Cute </paragraph>
<quote><focus/> Something else </quote>
```

I want the result to be 
```
<paragraph>x<cursor/></paragraph>
<quote> Something else </quote>
```
instead of 
```
<quote>xSomething else </quote>
```

I think it is surprising users when they find the next text are going to the next block, especially when the next block is a table cell or a list (or void blocks).

I think it is better for users to keep the insert text under the same block type, rather than using the next block.

**Comment by ianstormtaylor** [comment]
I'd actually like to have it the other way. Google Docs, Dropbox Paper and Quill.js all do it the other way.

**Comment by ianstormtaylor** [comment]
@zhujinxuan you make good points!

In that case, I feel like we should instead get rid of the "hanging" concept altogether, since you're right it does add lots of complexity. The only way I can think to do this though is to fix triple-click to nudge the selection back into the previous block. (Since that's the main source of hanging selections.)

Do you know if that's possible?

**Comment by zhujinxuan** [comment]
@ianstormtaylor Shall we remove block when we hit delete in a whole selected paragraph like this:

```
<paragraph><anchor/>aaaaaa<focus/></paragraph>
```
If we can delete that paragraph, then I think we can do some methods to avoid the triple selection in Slate.  (I think we can use range.normalize() to solve the problem)

**Comment by ianstormtaylor** [comment]
I think **1.** is okay.

I think **2.** is okay, but it can't be always when hanging. Instead, we need to be able to only nudge when a triple-click has occurred. Because otherwise you won't be able to select "through" that invalid selection if it auto-normalizes you.

For the inserting table case, I think that's not an issue? I think if people are using the "replace or insert" logic they'll be fine.

**Comment by zhujinxuan** [comment]
@ianstormtaylor Great. But then I just come up with another problem:
```
<paragraph><anchor/>aaaaaa<focus/></paragraph>
```
press enter
1. with normalize range
```
<paragraph></paragraph>
<paragraph><cursor/></paragraph>
```
2. without normalize range
```
<paragraph><cursor/></paragraph>
```
Do you think the double paragraph blocks are good?

**Comment by ianstormtaylor** [comment]
Yeah I think it’s fine. It’s kinda unclear what enter should do there, and using delete instead is more intuitive I think.

**Comment by zhujinxuan** [comment]
@ianstormtaylor Great. I will close this PR, and submit another PR with nudging triple-clicks and fixed insertTextAtRange

**Pull request closed without merging**